### PR TITLE
fix(markdown): do not break before special prefix

### LINKED ISSUE #3168: Markdown: unstable lists
**Prettier 1.8.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBiVACAygVwA54QBO8AJhgEo4A2cAzgDpRMC0GA2lBC6QJZ0BDAEa1SLeHRh0AugAoA9HgFgA1gIDm9efWq9YLPNRzq9LAFb0Y80hDB15RGlq49+w0eMt0AdAFtSAJQYbHyC1NQQAO4Yoe5w5BJS3qwcLgBmtjh08Z6SMgpKqhpaOnowBkYmUOaW1rb2jrT26ZnZYok+-kEhbuFRGBlgWfEYHcnVqdy8pAgwvGAC1OK8MLRyispqmval+obGphaSdXYOTs1TM7Dzi8urcH6BwTG9EdHTszfUoytN42zsABui2mLDgAA88HAwDB1oUtiU6Lo9pVDrUbKdGlpgboxBCoTDHt0MAg6DgiHASZDoTAMPwMAtwiNIEQKTDqABPcZMdDYfCEEgjKhNJhMABUk1cghEOQ6cM2xR2SLKFQO1SOVgxDXO8hcsRl7S8RKYGGesT60X1olGRvFksGw0NeXlRW22mVKLVNWOWrOTV13AdbVySS6JrNr36QZGYztnEun0ZdzWBQVbt25X2VW9mvqfucCeuSbm92NUFNPTCbzpVzmjJ+9x8cZxoPxNJdCKVyMzqPV6LzWPsLbx1MJXWC4dJ5MpbZhdLoDMW1pZbNWXJAABoQBA8HNoHRkKABJJkGlFtkt0IiJs4DAsIU9OpkDBHHAtxjT+e3yAzHRwQAha9VFvLABF8OAABk9DgT9qAvbccBgPBEIAJlg+ClCIbIiGQEBfAEIgVBsSIoE3EA8CIMoAHVphgAALZAAA4AAYt2yXxeGfV82MfWgAEUcAgeB0O-GBhBo0h6OQFCtxfAReGRdQAGEIF8fDcK4KAYK3YYABVhAPJAzzguAAF9TKAA)
```sh
--parser markdown
```

**Input:**
```markdown
## Supported Rules

- [no-disabled-tests](/packages/eslint-plugin-jest/docs/rules/no-disabled-tests.md) - disallow disabled tests.
- [no-focused-tests](/packages/eslint-plugin-jest/docs/rules/no-focused-tests.md) - disallow focused tests.
- [no-identical-title](/packages/eslint-plugin-jest/docs/rules/no-identical-title.md) - disallow identical titles.
- [valid-expect](/packages/eslint-plugin-jest/docs/rules/valid-expect.md) - ensure expect is called correctly.

## Supported Rules

* [no-disabled-tests](/packages/eslint-plugin-jest/docs/rules/no-disabled-tests.md)
  - disallow disabled tests.
* [no-focused-tests](/packages/eslint-plugin-jest/docs/rules/no-focused-tests.md)
  - disallow focused tests.
* [no-identical-title](/packages/eslint-plugin-jest/docs/rules/no-identical-title.md)
  - disallow identical titles.
* [valid-expect](/packages/eslint-plugin-jest/docs/rules/valid-expect.md) -
  ensure expect is called correctly.
```

**Output:**
```markdown
## Supported Rules

* [no-disabled-tests](/packages/eslint-plugin-jest/docs/rules/no-disabled-tests.md)
  - disallow disabled tests.
* [no-focused-tests](/packages/eslint-plugin-jest/docs/rules/no-focused-tests.md)
  - disallow focused tests.
* [no-identical-title](/packages/eslint-plugin-jest/docs/rules/no-identical-title.md)
  - disallow identical titles.
* [valid-expect](/packages/eslint-plugin-jest/docs/rules/valid-expect.md) -
  ensure expect is called correctly.

## Supported Rules

* [no-disabled-tests](/packages/eslint-plugin-jest/docs/rules/no-disabled-tests.md)
  * disallow disabled tests.
* [no-focused-tests](/packages/eslint-plugin-jest/docs/rules/no-focused-tests.md)
  * disallow focused tests.
* [no-identical-title](/packages/eslint-plugin-jest/docs/rules/no-identical-title.md)
  * disallow identical titles.
* [valid-expect](/packages/eslint-plugin-jest/docs/rules/valid-expect.md) -
  ensure expect is called correctly.

```

**Expected behavior:**
No ~change~ breaking of input, as it creates sublists

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #3168

The unstable is caused by the AST change, should be stable now.

### CODE REVIEW

File: tests/markdown_paragraph/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -42,3 +42,51 @@ This is a long long long long long long long long long long long long long long
...
+  disallow focused tests.
+* [no-identical-title](/packages/eslint-plugin-jest/docs/rules/no-identical-title.md) -
+  disallow identical titles.
+* [valid-expect](/packages/eslint-plugin-jest/docs/rules/valid-expect.md) -
+  ensure expect is called correctly.
+
+## Supported Rules
+
+* [no-disabled-tests](/packages/eslint-plugin-jest/docs/rules/no-disabled-tests.md)
+  * disallow disabled tests.
```
Comment (azz) [inline tests/markdown_paragraph/__snapshots__/jsfmt.spec.js.snap]: Should the nested list use `-` instead?

Comment (ikatyang) [reply tests/markdown_paragraph/__snapshots__/jsfmt.spec.js.snap]: If we use `-` for nested list, how the third-level list style should be? maybe opening an new issue for discussion would be better.

Comment (azz) [reply tests/markdown_paragraph/__snapshots__/jsfmt.spec.js.snap]: Sure - #3174

File: src/printer-markdown.js [hunk 2]
```diff
@@ -74,10 +74,20 @@ function genericPrint(path, options, print) {
...
             .replace(/(^|[^\\])\*/g, "$1\\*") // escape all unescaped `*` and `_`
             .replace(/\b(^|[^\\])_\b/g, "$1\\_"); // `1_2_3` is not considered emphasis
-    case "whitespace":
+    case "whitespace": {
+      const parentNode = path.getParentNode();
+      const index = parentNode.children.indexOf(node);
+      const nextNode = parentNode.children[index + 1];
+
+      // special prefix that may cause different meaning
+      if (nextNode && nextNode.value.match(/^(#{1,6}|>+)|^(-|\+|\*)$/)) {
```
Comment (lydell) [inline src/printer-markdown.js]: This could have been written as:

```js
if (nextNode && /^(#{1,6}|>)|^[-+*])$/.test(nextNode.value)) {
```

// Mr. Regex

**Pull request merged**